File: lib/llmClient.ts
import { exec } from 'child_process';
import { promisify } from 'util';
import path from 'path';
import fs from 'fs';

const execPromise = promisify(exec);

interface LlmClientOptions {
  pythonPath?: string;
  scriptPath?: string;
  maxRetries?: number;
}

export interface LlmResponse {
  answer: string;
  error?: string;
}

/**
 * Client for interacting with the Python LLM service
 */
export class LlmClient {
  private pythonPath: string;
  private scriptPath: string = ''; // Initialize to empty string
  private maxRetries: number;

  constructor(options: LlmClientOptions = {}) {
    // Always use python3 as our diagnostic showed it's available and working
    this.pythonPath = options.pythonPath || 'python3';
    
    // Search for the script in multiple locations
    const possiblePaths = [
      // Path provided in options
      options.scriptPath,
      // Path in our lib directory (local copy)
      path.resolve(process.cwd(), 'lib/llm_client.py'),
      // Original path in avos-bot-clean
      path.resolve(process.cwd(), '../avos-bot-clean/src/llm_client.py'),
      // Other possible locations
      path.resolve(process.cwd(), 'src/llm_client.py')
    ].filter(Boolean) as string[];
    
    // Find the first path that exists
    let scriptFound = false;
    for (const possiblePath of possiblePaths) {
      if (fs.existsSync(possiblePath)) {
        this.scriptPath = possiblePath;
        console.log(`Using LLM script at: ${this.scriptPath}`);
        scriptFound = true;
        break;
      }
    }
    
    // If no script was found, use the default path and warn
    if (!scriptFound) {
      this.scriptPath = path.resolve(process.cwd(), 'lib/llm_client.py');
      console.warn(`LLM script not found. Using default path: ${this.scriptPath}`);
    }
    
    this.maxRetries = options.maxRetries || 2;
  }

  /**
   * Get a response from the LLM based on the query and context
   */
  async getResponse(
    query: string, 
    context: string = '', 
    systemPrompt: string = '',
    conversationHistory: any[] = [],
    dbData: any = null
  ): Promise<LlmResponse> {
    // Prepare parameters
    const escapedQuery = JSON.stringify(query);
    const escapedContext = JSON.stringify(context);
    const escapedSystemPrompt = JSON.stringify(systemPrompt);
    const escapedConversationHistory = JSON.stringify(conversationHistory);
    
    // Construct command with properly escaped parameters
    let command = `${this.pythonPath} "${this.scriptPath}" ${escapedQuery} ${escapedContext} ${escapedSystemPrompt} ${escapedConversationHistory}`;
    
    // Add DB data if provided, making sure it's valid JSON
    if (dbData) {
      try {
        // Test that we can stringify it without errors
        const escapedDbData = JSON.stringify(dbData);
        command += ` ${escapedDbData}`;
      } catch (jsonError) {
        console.warn('Error stringifying DB data:', jsonError);
        // Don't add invalid data to the command
      }
    }
    
    try {
      // First try the normal LLM approach with retries
      const llmResponse = await this.tryLlmWithRetries(command, query);
      return llmResponse;
    } catch (error) {
      console.error('All LLM attempts failed, using guaranteed fallback response:', error);
      
      // If everything fails, use this direct in-code fallback
      // This doesn't rely on any external process or script
      return this.generateDirectFallbackResponse(query, context);
    }
  }
  
  // New method to handle retries
  private async tryLlmWithRetries(command: string, query: string): Promise<LlmResponse> {
    // Add debug logging for the command
    console.log('\n==== PYTHON COMMAND ====');
    console.log(command.substring(0, 150) + '...');
    console.log('=======================\n');
    
    // Execute with retries
    let attempt = 0;
    let lastError: Error | null = null;
    
    while (attempt < this.maxRetries) {
      try {
        console.log(`Executing LLM command (attempt ${attempt + 1}/${this.maxRetries})`);
        
        // Log timestamp for debugging
        console.log(`[${new Date().toISOString()}] Starting Python execution`);
        
        const { stdout, stderr } = await execPromise(command);
        
        // Debug logs for the output
        console.log(`\n==== PYTHON EXECUTION RESULT ====`);
        console.log(`STDERR (first 200 chars): ${stderr ? stderr.substring(0, 200) : 'None'}`);
        console.log(`STDOUT (first 100 chars): ${stdout ? stdout.substring(0, 100) + '...' : 'Empty response'}`);
        console.log(`STDOUT Content-Type indicators: HTML=${stdout.includes('<!DOCTYPE') || stdout.includes('<html>')}, JSON=${stdout.includes('{') && stdout.includes('}')}`);
        console.log(`=======================\n`);
        
        if (stderr) {
          console.warn(`LLM client warning: ${stderr}`);
        }
        
        if (!stdout || stdout.trim() === '') {
          throw new Error('Empty response from LLM client');
        }
        
        // Check if the response contains HTML markup which would indicate an error page
        if (stdout.includes('<!DOCTYPE') || stdout.includes('<html') || stdout.includes('</html>')) {
          console.error('Received HTML response instead of text:', stdout.substring(0, 100) + '...');
          throw new Error('Received HTML response instead of text');
        }
        
        // Sanitize the response to remove any HTML-like content that might cause issues
        const sanitizedResponse = stdout.replace(/<[^>]*>/g, '');
        
        return { answer: sanitizedResponse.trim() };
      } catch (error) {
        lastError = error as Error;
        console.error(`LLM client error (attempt ${attempt + 1}/${this.maxRetries}):`, error);
        attempt++;
      }
    }
    
    // If all attempts fail, try the fallback
    try {
      // Call the LLM script with a special flag to use simulated response
      console.log('Falling back to simulated response');
      console.log(`[${new Date().toISOString()}] Starting fallback execution`);
      
      const fallbackCommand = `${this.pythonPath} "${this.scriptPath}" ${JSON.stringify(query)}`;
      console.log(`Fallback command: ${fallbackCommand.substring(0, 100)}...`);
      
      const { stdout, stderr } = await execPromise(fallbackCommand);
      
      // Debug logs for the fallback output
      console.log(`\n==== PYTHON FALLBACK RESULT ====`);
      console.log(`STDERR: ${stderr ? stderr.substring(0, 200) : 'None'}`);
      console.log(`STDOUT (first 100 chars): ${stdout ? stdout.substring(0, 100) + '...' : 'Empty response'}`);
      console.log(`=======================\n`);
      
      if (stderr) {
        console.warn(`LLM client fallback warning: ${stderr}`);
      }
      
      // If fallback returns empty response, throw error to use the direct fallback
      if (!stdout || stdout.trim() === '') {
        throw new Error('Empty response from LLM fallback');
      }
      
      // Check for HTML in fallback response too
      if (stdout.includes('<!DOCTYPE') || stdout.includes('<html') || stdout.includes('</html>')) {
        throw new Error('Received HTML in fallback response');
      }
      
      // Sanitize the fallback response
      const sanitizedResponse = stdout.replace(/<[^>]*>/g, '');
      
      return { answer: sanitizedResponse.trim() };
    } catch (fallbackError) {
      console.error('Script fallback also failed:', fallbackError);
      
      // Log that we're using direct in-code fallback
      console.log('All Python attempts failed, using direct in-code fallback');
      
      throw fallbackError; // Let the parent method handle with direct fallback
    }
  }

  // New method for direct in-code fallback that doesn't rely on Python
  private generateDirectFallbackResponse(query: string, context: string = ''): LlmResponse {
    console.log('Using direct in-code fallback response');
    
    // Simple knowledge base for common queries
    const knowledgeBase: Record<string, string> = {
      'avos': 'AVOS (Autonomous Vehicle Operating System) is NVIDIA\'s comprehensive software stack designed for autonomous vehicles. It provides a flexible, scalable platform that integrates perception, planning, and control systems necessary for self-driving capabilities.',
      'drive': 'NVIDIA DRIVE is a platform that uses AVOS and is designed for developing autonomous vehicles. It includes both hardware (like the DRIVE AGX Orin system-on-a-chip) and software components that work together to enable self-driving capabilities.',
      'driveos': 'DriveOS is the operating system layer of NVIDIA\'s autonomous vehicle software stack. It provides a foundation for running autonomous driving applications, managing hardware resources, and ensuring real-time performance for critical driving functions.',
      'features': 'AVOS includes many features such as:\n- Sensor fusion for combining data from cameras, radar, and lidar\n- Perception systems for object detection and classification\n- Planning and decision-making algorithms\n- Control systems for vehicle operation\n- Simulation capabilities for testing and validation\n- Over-the-air update functionality',
      'help': 'You can ask me questions about AVOS (Autonomous Vehicle Operating System) developed by NVIDIA. I can provide information about its features, capabilities, architecture, and usage. For more detailed searches, try using "Go Search" before your query.',
    };
    
    // Convert query to lowercase for matching
    const queryLower = query.toLowerCase();
    
    // Default response if no match is found
    let response = "I apologize, but I'm currently experiencing technical difficulties connecting to the knowledge base. Here's what I can tell you though: AVOS (Autonomous Vehicle Operating System) is NVIDIA's comprehensive platform for autonomous vehicles. It integrates perception, planning, and control capabilities.";
    
    // Check for exact matches or keywords in the knowledge base
    for (const [key, value] of Object.entries(knowledgeBase)) {
      if (queryLower.includes(key)) {
        response = value;
        break;
      }
    }
    
    // If we have context, use it to enhance the response
    if (context && context.length > 0) {
      response += "\n\nBased on the available information, I can also add: " + context.substring(0, 500);
      if (context.length > 500) {
        response += "...";
      }
    }
    
    // Add a disclaimer
    response += "\n\n**_NOTE:_** I'm currently operating in a limited capacity due to connectivity issues. For more complete answers, please try again later.";
    
    return { 
      answer: response,
      error: "Using direct fallback response due to service unavailability"
    };
  }
  
  /**
   * Format content for optimal LLM processing
   */
  formatContentForLlm(content: any[], maxTokens: number = 6000): string {
    // Basic implementation - can be enhanced with more sophisticated chunking
    let formattedContent = '';
    
    if (Array.isArray(content)) {
      // Format array of content items
      for (const item of content) {
        if (typeof item === 'object') {
          // Format page content
          if (item.pageTitle && item.content) {
            formattedContent += `# ${item.pageTitle}\n\n`;
            
            // Process content to specially mark code blocks and limit to relevance
            const extractedContent = extractRelevantContent(item.content, 800);
            const processedContent = this.processContentWithCodeBlocks(extractedContent);
            formattedContent += `${processedContent}\n\n`;
          }
          
          // Add code snippets if available
          if (item.extractedElements) {
            const codeSnippets = item.extractedElements.filter((el: any) => 
              el.type === 'code' || el.type === 'code-block'
            );
            
            if (codeSnippets.length > 0) {
              formattedContent += "## Code Snippets\n\n";
              for (const snippet of codeSnippets.slice(0, 3)) {
                // Add special markers to indicate exact code blocks
                formattedContent += `<EXACT_CODE_BLOCK>\n\`\`\`\n${snippet.content}\n\`\`\`\n</EXACT_CODE_BLOCK>\n\n`;
              }
            }
          }
        } else if (typeof item === 'string') {
          formattedContent += `${item}\n\n`;
        }
      }
    } else if (typeof content === 'object') {
      // Single content object
      formattedContent = JSON.stringify(content);
    } else {
      // String or other type
      formattedContent = String(content);
    }
    
    // Simple truncation to keep within token limits
    // This is a very rough approximation - about 4 chars per token
    if (formattedContent.length > maxTokens * 4) {
      formattedContent = formattedContent.substring(0, maxTokens * 4) + '...';
    }
    
    return formattedContent;
  }
  
  /**
   * Process content to specially mark code blocks for preservation
   */
  private processContentWithCodeBlocks(content: string): string {
    if (!content) return '';
    
    const lines = content.split('\n');
    let processedLines: string[] = [];
    let inCodeBlock = false;
    let codeBlockLang = '';
    let codeBlockContent: string[] = [];
    
    for (const line of lines) {
      // Check for code block start
      if (line.trim().startsWith('```') && !inCodeBlock) {
        inCodeBlock = true;
        codeBlockLang = line.trim().replace('```', '').trim();
        codeBlockContent = [line]; // Start collecting code block
      } 
      // Check for code block end
      else if (line.trim() === '```' && inCodeBlock) {
        inCodeBlock = false;
        codeBlockContent.push(line); // Add closing marker
        
        // Add special tags around code block
        processedLines.push('<EXACT_CODE_BLOCK>');
        processedLines.push(...codeBlockContent);
        processedLines.push('</EXACT_CODE_BLOCK>');
        
        codeBlockContent = []; // Reset code block content
      } 
      // Inside code block - collect content
      else if (inCodeBlock) {
        codeBlockContent.push(line);
      } 
      // Regular content
      else {
        processedLines.push(line);
      }
    }
    
    // In case we ended with an unclosed code block
    if (inCodeBlock && codeBlockContent.length > 0) {
      processedLines.push('<EXACT_CODE_BLOCK>');
      processedLines.push(...codeBlockContent);
      processedLines.push('```'); // Add closing marker
      processedLines.push('</EXACT_CODE_BLOCK>');
    }
    
    return processedLines.join('\n');
  }
}

// Helper to extract most relevant content by truncating long text
function extractRelevantContent(text: string, maxChars: number = 800): string {
  if (text.length <= maxChars) return text;
  
  // Split by newlines to preserve structure
  const lines = text.split('\n');
  let result = '';
  let charCount = 0;
  
  for (const line of lines) {
    if (charCount + line.length > maxChars) {
      // Check if we can fit at least part of this line
      const remainingChars = maxChars - charCount;
      if (remainingChars > 20) { // Only add if we can fit a meaningful chunk
        result += line.substring(0, remainingChars) + '...';
      }
      break;
    }
    
    result += line + '\n';
    charCount += line.length + 1; // +1 for newline
  }
  
  return result;
}

// Create singleton instance
const llmClient = new LlmClient();
export default llmClient; 

File: lib/llm_client.py
#!/usr/bin/env python3

import os
import time
import json
import sys
import base64
import re
from pathlib import Path

# Handle import errors gracefully
try:
    import requests
    from openai import AzureOpenAI
    DEPENDENCIES_INSTALLED = True
except ImportError:
    DEPENDENCIES_INSTALLED = False
    print("Warning: Required dependencies not installed. Using simulated responses only.", file=sys.stderr)
    print("To install dependencies: pip install requests openai", file=sys.stderr)
except Exception as e:
    DEPENDENCIES_INSTALLED = False
    print(f"Unexpected error importing dependencies: {e}", file=sys.stderr)
    print("Using simulated responses only.", file=sys.stderr)

# Safety wrapper to ensure we always return a valid response
def safe_response(func):
    def wrapper(*args, **kwargs):
        try:
            result = func(*args, **kwargs)
            # Ensure result is a string
            if not isinstance(result, str):
                return "Error: Invalid response type from function"
            return result
        except Exception as e:
            print(f"Error in {func.__name__}: {e}", file=sys.stderr)
            return f"I apologize, but I encountered an error while processing your request. Please try again with a simpler query."
    return wrapper

# Read prompts from the JS file
@safe_response
def get_default_system_prompt():
    """Read the default system prompt from prompts.js config file."""
    try:
        script_dir = os.path.dirname(os.path.abspath(__file__))
        prompts_path = os.path.join(script_dir, "config", "prompts.js")
        
        if os.path.exists(prompts_path):
            with open(prompts_path, 'r') as f:
                content = f.read()
                # Simple extraction of the DEFAULT_LLM_PROMPT from the JS file
                start = content.find('DEFAULT_LLM_PROMPT = `') + len('DEFAULT_LLM_PROMPT = `')
                end = content.find('`;', start)
                if start > 0 and end > 0:
                    return content[start:end]
        
        # Fallback prompt if file not found or parsing fails
        return "You are a helpful AI assistant specialized in AVOS (Autonomous Vehicle Operating System) developed by NVIDIA. Provide accurate and helpful information about AVOS features, capabilities, and usage. If you don't know something, be honest about it. Please give small and precise answers. Don't give out of context answers. Please give answers in bullet points."
    except Exception as e:
        print(f"Error loading system prompt: {e}", file=sys.stderr)
        return "You are a helpful AI assistant specialized in AVOS (Autonomous Vehicle Operating System) developed by NVIDIA. Provide accurate and helpful information about AVOS features, capabilities, and usage. If you don't know something, be honest about it."

def basic_auth(p_client_id, p_client_secret):
    """Create Basic Auth header value."""
    token = base64.b64encode(f"{p_client_id}:{p_client_secret}".encode('utf-8')).decode("ascii")
    return f'Basic {token}'

def get_oauth_token(p_token_url, p_client_id, p_client_secret, p_scope):
    """Get OAuth token from server or cache."""
    if not DEPENDENCIES_INSTALLED:
        return None
        
    file_name = "py_llm_oauth_token.json"
    try:
        base_path = Path(__file__).parent
        file_path = Path.joinpath(base_path, file_name)
    except Exception as e:
        print(f"Error occurred while setting file path: {e}", file=sys.stderr)
        return None

    try:
        # Check if the token is cached
        if os.path.exists(file_path):
            with open(file_path, "r") as f:
                token = json.load(f)
                
                # Check if the token is expired
                if "expires_in" in token and "created_at" in token:
                    expiry_time = token["created_at"] + token["expires_in"]
                    if time.time() < expiry_time:
                        return token["access_token"]
        
        # Get a new token from the OAuth server
        payload = "grant_type=client_credentials&scope=" + p_scope
        headers = {
            "Content-Type": "application/x-www-form-urlencoded",
            "Authorization": basic_auth(p_client_id, p_client_secret)
        }
        
        response = requests.request("POST", p_token_url, headers=headers, data=payload)
        token = response.json()
        
        # Add creation timestamp
        token["created_at"] = time.time()
        
        with open(file_path, "w") as f:
            json.dump(token, f)
            
        return token["access_token"]
    except Exception as e:
        print(f"Error occurred while getting OAuth token: {e}", file=sys.stderr)
        return None

def get_azure_client():
    """Get authenticated Azure OpenAI client."""
    if not DEPENDENCIES_INSTALLED:
        return None
        
    try:
        # Azure OpenAI credentials
        client_id = "nvssa-prd-_lwo4JYj0iZDQaQhmQE64W5gl25gqmvjnBjvY-6YSuU"
        client_secret = "ssap-ryqWx5WbHYNulr9TPcA"
        token_url = "https://5kbfxgaqc3xgz8nhid1x1r8cfestoypn-trofuum-oc.ssa.nvidia.com/token"
        scope = "azureopenai-readwrite"

        token = get_oauth_token(token_url, client_id, client_secret, scope)
        if not token:
            return None
            
        api_base = "https://prod.api.nvidia.com/llm/v1/azure/openai"
        api_version = "2023-12-01-preview"

        client = AzureOpenAI(api_key=token, api_version=api_version, base_url=api_base)
        return client
    except Exception as e:
        print(f"Error creating Azure OpenAI client: {e}", file=sys.stderr)
        return None

def format_db_data_for_llm(db_data):
    """
    Format data retrieved from database to be included in LLM context.
    This helps provide structured information to the LLM while preserving
    code blocks exactly as they appear.
    """
    if not db_data:
        return ""
    
    # Special header to signal the importance of preserving code blocks    
    formatted_data = """
IMPORTANT: The following information contains code blocks that must be preserved EXACTLY as shown,
with their original formatting, indentation, comments, and whitespace.
===== DATABASE INFORMATION START =====

"""
        
    if isinstance(db_data, list):
        # Format list of records
        for i, item in enumerate(db_data):
            if isinstance(item, dict):
                # Add page title if available
                if "pageTitle" in item:
                    formatted_data += f"## {item.get('pageTitle')}\n\n"
                
                # Add content with special handling for code blocks
                if "content" in item:
                    content = item.get("content", "")
                    # Try to detect and specially mark code blocks
                    lines = content.split('\n')
                    in_code_block = False
                    code_block_lang = ""
                    
                    for line in lines:
                        # Detect code block markers in Markdown or HTML
                        if line.strip().startswith("```") and not in_code_block:
                            in_code_block = True
                            # Try to extract language info
                            code_block_lang = line.strip().replace("```", "").strip()
                            # Add special marker for code block start
                            formatted_data += f"\n<CODE_BLOCK_START lang=\"{code_block_lang}\">\n"
                            formatted_data += line + "\n"
                        elif line.strip() == "```" and in_code_block:
                            in_code_block = False
                            formatted_data += line + "\n"
                            # Add special marker for code block end
                            formatted_data += "<CODE_BLOCK_END>\n\n"
                        else:
                            formatted_data += line + "\n"
                    
                    formatted_data += "\n"
                
                # Add other fields from the item
                for key, value in item.items():
                    if key not in ["pageTitle", "content"]:
                        formatted_data += f"- {key}: {value}\n"
                
                formatted_data += "\n---\n\n"
            else:
                formatted_data += f"Item {i+1}:\n{item}\n\n"
    elif isinstance(db_data, dict):
        # Format dictionary
        formatted_data += "Retrieved information:\n\n"
        for key, value in db_data.items():
            formatted_data += f"- {key}: {value}\n"
    else:
        # Just convert to string if not a recognized format
        formatted_data += f"Retrieved information:\n\n{db_data}"
        
    formatted_data += "\n===== DATABASE INFORMATION END =====\n"
    
    # Add instruction for handling code blocks
    formatted_data += """
REMEMBER: When including code blocks in your response, reproduce them EXACTLY as shown above,
with the same formatting, indentation, comments, and whitespace. Do not modify any code.
"""
    
    return formatted_data

def enhance_response_formatting(response):
    """
    Enhance the response from LLM with proper formatting based on content type.
    - Ensures code blocks are properly formatted
    - Detects and formats important information
    - Applies Markdown formatting
    """
    # Make sure code blocks are properly formatted
    # Look for code that might not be in code blocks
    if "```" not in response and re.search(r'(import\s+[\w\.]+|def\s+\w+\(|class\s+\w+\(|function\s+\w+\()', response):
        # This looks like code but isn't in a code block
        lines = response.split("\n")
        in_code_block = False
        formatted_lines = []
        
        for line in lines:
            if re.match(r'^(import\s+[\w\.]+|def\s+\w+\(|class\s+\w+\(|function\s+\w+\()', line.strip()) and not in_code_block:
                formatted_lines.append("```python")
                in_code_block = True
            elif in_code_block and line.strip() == "" and len(formatted_lines) > 0:
                formatted_lines.append("```")
                in_code_block = False
            
            formatted_lines.append(line)
            
        if in_code_block:
            formatted_lines.append("```")
            
        response = "\n".join(formatted_lines)
    
    # Look for important keywords and highlight them
    important_keywords = ["IMPORTANT", "WARNING", "CRITICAL", "NOTE", "CAUTION"]
    for keyword in important_keywords:
        pattern = rf'({keyword}:)(.*?)(\n|$)'
        replacement = r'**_\1_** \2\3'
        response = re.sub(pattern, replacement, response, flags=re.IGNORECASE)
    
    # Make sure bullet points are properly formatted
    response = re.sub(r'(?<!\n)^(\d+\.\s)', r'\n\1', response, flags=re.MULTILINE)
    response = re.sub(r'(?<!\n)^(-\s)', r'\n\1', response, flags=re.MULTILINE)
    
    return response

@safe_response
def get_simulated_response(prompt, context=""):
    """Provide simulated responses when API is unavailable."""
    # The simulated knowledge base now comes from config/prompts.js
    # This is a simplified version as fallback
    avos_knowledge = {
        'avos': 'AVOS (Autonomous Vehicle Operating System) is NVIDIA\'s comprehensive software stack designed for autonomous vehicles. It provides a flexible, scalable platform that integrates perception, planning, and control systems necessary for self-driving capabilities.',
        'drive': 'NVIDIA DRIVE is a platform that uses AVOS and is designed for developing autonomous vehicles. It includes both hardware (like the DRIVE AGX Orin system-on-a-chip) and software components that work together to enable self-driving capabilities.',
        'driveos': 'DriveOS is the operating system layer of NVIDIA\'s autonomous vehicle software stack. It provides a foundation for running autonomous driving applications, managing hardware resources, and ensuring real-time performance for critical driving functions.',
        'ndas': 'NDAS (NVIDIA Data Annotation System) is a tool for labeling and annotating sensor data collected from vehicles. It helps create training datasets for machine learning models used in autonomous driving systems.',
        'feature': 'AVOS includes many features such as:\n- Sensor fusion for combining data from cameras, radar, and lidar\n- Perception systems for object detection and classification\n- Planning and decision-making algorithms\n- Control systems for vehicle operation\n- Simulation capabilities for testing and validation\n- Over-the-air update functionality',
        'dtsi': 'In the context of DriveOS, a DTSI (Device Tree Source Include) file is used to describe hardware components and their properties. The "startupcmd" DTSI file specifically contains commands that are executed during system startup to configure hardware components and initialize services.',
        'steps': 'To integrate DriveOS changes into NDAS, you would typically follow these steps:\n1. Develop and test your changes in a DriveOS development environment\n2. Document the changes thoroughly\n3. Submit the changes through the code review process\n4. Work with the NDAS team to integrate and test the changes\n5. Monitor and validate the integration through regression testing',
        'secpolicy': 'AVOS Customizations for Security Policy (SecPolicy) include configurations for mounting policies and folder hierarchy support. SecPolicy files are available for both debug and production environments, typically named policy_debug_orin_gos_vm_safety.txt and policy_prod_orin_gos_vm_safety.txt respectively. The security policy enforces folder hierarchy and mounting restrictions.',
    }
    
    try:
        # Convert prompt to lowercase for easier matching
        prompt_lower = prompt.lower()
        
        # Default response
        response = "I don't have specific information about that aspect of AVOS. Could you ask something more general about AVOS capabilities or features?"
        
        # Try to find relevant information in our knowledge base
        for keyword, info in avos_knowledge.items():
            if keyword.lower() in prompt_lower:
                response = info
                break
        
        # Use context if available and we don't have a good response yet
        if context and "I don't have specific information" in response:
            response = f"Based on the available information: {context}\n\nHowever, please note that this information may be limited. For more detailed information, please consult the official NVIDIA AVOS documentation."
        
        # Special case for questions about the relationship between DriveOS and NDAS
        if 'driveos' in prompt_lower and 'ndas' in prompt_lower:
            response = avos_knowledge['steps']
        
        # Special case for questions about DTSI files
        if 'dtsi' in prompt_lower and 'startupcmd' in prompt_lower:
            response = avos_knowledge['dtsi']
        
        # Special case for security policy questions
        if 'secpolicy' in prompt_lower or ('security' in prompt_lower and 'policy' in prompt_lower):
            response = avos_knowledge['secpolicy']
        
        return enhance_response_formatting(response)
    except Exception as e:
        print(f"Error in simulated response: {e}", file=sys.stderr)
        return "I apologize, but I'm having difficulty processing your request right now. Please try again with a simpler query."

@safe_response
def get_llm_response(prompt, context="", system_prompt="", conversation_history_json="[]", db_data=None):
    """Get response from LLM, falling back to simulated responses if unavailable."""
    try:
        # Format database data if provided
        if db_data:
            db_context = format_db_data_for_llm(db_data)
            if context:
                context = f"{context}\n\n{db_context}"
            else:
                context = db_context
        
        if DEPENDENCIES_INSTALLED:
            try:
                # Get the Azure OpenAI client
                client = get_azure_client()
                
                if client:
                    # Prepare messages
                    messages = []
                    
                    # Add system prompt if provided, otherwise use default
                    if system_prompt:
                        messages.append({"role": "system", "content": system_prompt})
                    else:
                        messages.append({
                            "role": "system", 
                            "content": get_default_system_prompt()
                        })
                    
                    # Add conversation history if provided
                    try:
                        conversation_history = json.loads(conversation_history_json)
                        if conversation_history and isinstance(conversation_history, list):
                            messages.extend(conversation_history)
                    except Exception as e:
                        print(f"Error parsing conversation history: {e}", file=sys.stderr)
                    
                    # Add context if provided
                    if context:
                        messages.append({"role": "user", "content": f"Context information: {context}"})
                    
                    # Enhance the prompt to encourage proper formatting in the response
                    formatting_guidance = """
When responding, please format your answer appropriately:
- Use code blocks with proper language specification for any code (```python, ```bash, etc.)
- Highlight IMPORTANT information in bold and italic format (**_important_**)
- Use proper Markdown for lists, headings, and other formatting
- If you're explaining steps, use numbered lists
- Use bullet points for feature lists
"""
                    enhanced_prompt = f"{prompt}\n\n{formatting_guidance}"
                    
                    # Add user prompt
                    messages.append({"role": "user", "content": enhanced_prompt})
                    
                    # Call the OpenAI Chat Completion
                    completion = client.chat.completions.create(
                        model="gpt-4o",
                        messages=messages,
                        max_tokens=1000
                    )
                    
                    # Get and enhance the response
                    response = completion.choices[0].message.content
                    return enhance_response_formatting(response)
            except Exception as e:
                print(f"Error calling LLM API: {e}", file=sys.stderr)
                print("Falling back to simulated response.", file=sys.stderr)
        
        # If dependencies aren't installed or API call failed, use simulated response
        return get_simulated_response(prompt, context)
    except Exception as e:
        print(f"Critical error in get_llm_response: {e}", file=sys.stderr)
        return "I apologize for the inconvenience, but I'm experiencing technical difficulties. Please try again later."

if __name__ == "__main__":
    # This script can be called from command line with arguments:
    # arg1: user prompt
    # arg2 (optional): context
    # arg3 (optional): system prompt
    # arg4 (optional): conversation history JSON
    # arg5 (optional): database data JSON
    
    try:
        if len(sys.argv) < 2:
            print("Usage: python llm_client.py \"prompt\" [\"context\"] [\"system_prompt\"] [\"conversation_history_json\"] [\"db_data_json\"]")
            sys.exit(1)
        
        prompt = sys.argv[1]
        context = sys.argv[2] if len(sys.argv) > 2 else ""
        system_prompt = sys.argv[3] if len(sys.argv) > 3 else ""
        conversation_history_json = sys.argv[4] if len(sys.argv) > 4 else "[]"
        db_data_json = sys.argv[5] if len(sys.argv) > 5 else None
        
        db_data = None
        if db_data_json:
            try:
                db_data = json.loads(db_data_json)
            except Exception as e:
                print(f"Error parsing database data: {e}", file=sys.stderr)
        
        # Get the response in a try-catch block to handle any errors
        try:
            response = get_llm_response(prompt, context, system_prompt, conversation_history_json, db_data)
            
            # Ensure we're not outputting invalid content
            if not response or not isinstance(response, str):
                response = "Error: Invalid response from LLM"
                
            # Strip any HTML-like content that could cause JSON parsing errors
            response = re.sub(r'<[^>]*>', '', response)
            
            # Output the sanitized response
            print(response)
        except Exception as e:
            # If there's any error in LLM processing, return a safe error message
            print(f"I apologize, but I encountered an error processing your request. Please try again with a simpler query.")
            print(f"Error details: {str(e)}", file=sys.stderr)
    except Exception as e:
        # Catch-all for any other errors
        print("I apologize, but there was an error processing your request.")
        print(f"Critical error: {str(e)}", file=sys.stderr) 